File: Trees_Graphs/go/successor.go
package main
import (
  "types"
  "fmt"
)


func getNext(node *types.TreeNodeWithParent) *types.TreeNodeWithParent {
  if node.LeftChild != nil {
    return getNext(node.LeftChild)
  }

  return node
}


func successor(node *types.TreeNodeWithParent) *types.TreeNodeWithParent {
  if node.RightChild == nil {
    parent := node.Parent
    for parent.Value < node.Value {
      parent = parent.Parent

      if parent == nil {
        return &types.TreeNodeWithParent{}
      }
    }

    return parent
  }

  return getNext(node.RightChild)
}


func main() {
  root := types.TreeNodeWithParent{Value: 4}

  secondLevelLeftChild := types.TreeNodeWithParent{Value: 2, Parent: &root}
  secondLevelRightChild := types.TreeNodeWithParent{Value: 6, Parent: &root}
  root.LeftChild = &secondLevelLeftChild
  root.RightChild = &secondLevelRightChild

  thirdLevelFirst := types.TreeNodeWithParent{Value: 1, Parent: &secondLevelLeftChild}
  thirdLevelSecond := types.TreeNodeWithParent{Value: 3, Parent: &secondLevelLeftChild}
  thirdLevelThird := types.TreeNodeWithParent{Value: 5, Parent: &secondLevelRightChild}
  thirdLevelFourth := types.TreeNodeWithParent{Value: 7, Parent: &secondLevelRightChild}

  secondLevelLeftChild.LeftChild = &thirdLevelFirst
  secondLevelLeftChild.RightChild = &thirdLevelSecond
  secondLevelRightChild.LeftChild = &thirdLevelThird
  secondLevelRightChild.RightChild = &thirdLevelFourth

  fmt.Println(successor(&thirdLevelFourth))
}

File: Trees_Graphs/python/successor.py
from Tree import *


def get_next(node):
    if node.left_child is not None:
        return get_next(node.left_child)

    return node


def successor(node):
    if node.right_child is None:
        parent = node.parent
        while parent.value < node.value:
            parent = parent.parent

            if parent is None:
                return None

        return parent

    return get_next(node.right_child)


def main():
    array = list(range(7))
    tree = Tree(array)
    tree.print_tree()

    print(successor(tree.root.right_child.right_child))


if __name__ == '__main__':
    main()
